src/NLog/Targets/WebServiceTarget.cs

```diff
@@ -199,6 +200,162 @@ public WebServiceTarget(string name) : this()
+                : logEvents.BucketSort(convetToParameterArrayDelegate, ArrayDeepEqualityComparer<object>.Default);
+            foreach (var bucket in parameterBuckets)
+            {
+                for (int i = 0; i < this.Parameters.Count; ++i)
+                {
+                    var param = this.Parameters[i];
+                    if (param.GroupLayoutActive)
+                    {
+                        bucket.Key[i] = ConvertParameterGroupValue(bucket.Value, param);
+                    }
+                }
+
+                if (bucket.Value.Count > 1)
+                {
+                    AsyncContinuation[] groupContinuations = new AsyncContinuation[bucket.Value.Count];
+                    for (int i = 0; i < groupContinuations.Length; ++i)
+                    {
+                        groupContinuations[i] = bucket.Value[i].Continuation;
+                    }
+                    AsyncContinuation groupCompleted = (ex) =>
+                    {
+                        for (int i = 0; i < groupContinuations.Length; ++i)
+                            try { groupContinuations[i].Invoke(ex); } catch { /* Nothing to do about it */ };
+                    };
+                    DoGroupInvokeAsync(headerValues, bucket.Key, groupCompleted);
+                }
+                else
+                {
+                    DoGroupInvokeAsync(headerValues, bucket.Key, bucket.Value[0].Continuation);
+                }
+            }
+        }
+        private SortHelpers.KeySelector<AsyncLogEventInfo, object[]> convetToParameterArrayDelegate;
+
+        class ArrayDeepEqualityComparer<TValue> : IEqualityComparer<TValue[]>
+        {
+            public static readonly ArrayDeepEqualityComparer<TValue> Default = new ArrayDeepEqualityComparer<TValue>();
+
+            public bool Equals(TValue[] x, TValue[] y)
+            {
+                if (x.Length != y.Length)
+                    return false;
+
+                object xval, yval;
+                for (int i = 0; i < x.Length; ++i)
+                {
+                    xval = x[i];
+                    yval = y[i];
+                    if (xval != null && yval != null)
+                    {
+                        if (!x[i].Equals(y[i]))
+                            return false;
+                    }
+                    else if (!ReferenceEquals(xval, yval))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            public int GetHashCode(TValue[] obj)
+            {
+                int hashCode = obj.Length.GetHashCode();
+                for (int i = 0; i < obj.Length; ++i)
+                {
+                    if (obj[i] != null)
+                        hashCode = hashCode ^ obj[i].GetHashCode();
```

Comment:
Well I find Visual Studio really slow and annoying, and RSharper really doesn't help in that area. But maybe I start using it when I have grown a little wiser :). I'm guessing the that the GetHashCode()-methods already uses a prime-number (not using the raw-values). I know that XOR will give many collisions if the parameter-array-values are all the same, but I'm guessing they will not be. And the GetHashCode() doesn't have to be that optimal as there most likely be one or two groups at a time.